### PR TITLE
add: tiled importer v.1.0 and renderer support for pixel

### LINKED ISSUE #16: TileD support
**Describe the solution you'd like**
Add support for TileD import and rendering capabilities

### PR DESCRIPTION
**Pull request by Homailot**
closes #16

### CODE REVIEW

**Comment by joafalves** [comment]
@Homailot consider merging feature changes on a single commit in future PRs.

File: .github/workflows/gradle.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# This workflow will build a Java project with Gradle and cache/restore any dependencies to improve the workflow execution time
```
Comment (joafalves) [inline .github/workflows/gradle.yml]: Remove this file (this can be added, but it should be a different PR)

File: build/tmp/jar/MANIFEST.MF [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Manifest-Version: 1.0
```
Comment (joafalves) [inline build/tmp/jar/MANIFEST.MF]: remove these build files (same for bellow items)

File: demos/tiled/src/main/java/org/pixel/demo/tiled/TiledDemo.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+public class TiledDemo extends PixelWindow {
+
+    private static final float CAMERA_SPEED = 100f;
+
+    private SpriteBatch spriteBatch;
+    private ContentManager contentManager;
+    private Camera2D gameCamera;
+    private static Color fillColor;
+    float posX, posY;
```
Comment (joafalves) [inline demos/tiled/src/main/java/org/pixel/demo/tiled/TiledDemo.java]: private?

File: demos/tiled/src/main/java/org/pixel/demo/tiled/TiledDemo.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        } else if (Keyboard.isKeyDown(KeyboardKey.D_3)) {
+            gameCamera.setZoom(3f);
+        }
+
+        titleFpsCounter.update(delta);
+    }
+
+    @Override
+    public void draw(DeltaTime delta) {
+        setBackgroundColor(fillColor);
```
Comment (joafalves) [inline demos/tiled/src/main/java/org/pixel/demo/tiled/TiledDemo.java]: setBackgroundColor shall only be called when the value changes. Btw, pixel wil automatically clear the screen at the begining of each frame (can be disabled on the window settings - requires manual call of "clear()" from `PixelWindow`).

File: demos/tiled/src/main/java/org/pixel/demo/tiled/TiledDemo.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+import org.pixel.input.keyboard.KeyboardKey;
+
+public class TiledDemo extends PixelWindow {
+
+    private static final float CAMERA_SPEED = 100f;
+
+    private SpriteBatch spriteBatch;
+    private ContentManager contentManager;
+    private Camera2D gameCamera;
+    private static Color fillColor;
```
Comment (joafalves) [inline demos/tiled/src/main/java/org/pixel/demo/tiled/TiledDemo.java]: Based on the utilization, this isn't required right? You can use Color.BLACK directly in `setBackgroundColor()` (on the load() function).

File: demos/tiled/src/main/java/org/pixel/demo/tiled/TiledDemo.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        spriteBatch.begin(gameCamera.getViewMatrix());
+
+        tileMapView.draw(tileMap, delta.getElapsedMs());
+
+        spriteBatch.end();
+    }
+
+    @Override
+    public void dispose() {
+        super.dispose();
```
Comment (joafalves) [inline demos/tiled/src/main/java/org/pixel/demo/tiled/TiledDemo.java]: `super.dispose()` shall be called last on this scenario.

File: extensions/ext-tiled/build.gradle [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+pitest {
+    threads = 4
+    targetClasses = ['org.pixel.ext.tiled.*']
+    excludedTestClasses = ['org.pixel.ext.tiled.**.*Integrated']
+    junit5PluginVersion = '0.12'
+}
+
+dependencies {
+    testImplementation('org.mockito:mockito-core:3.7.7')
```
Comment (joafalves) [inline extensions/ext-tiled/build.gradle]: `testImplementation` entries shall be added last (before`org.junit.jupiter:junit-jupiter-api:5.6.0` for example) .

File: extensions/ext-tiled/build.gradle [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    implementation 'org.junit.jupiter:junit-jupiter:5.7.0'
+    implementation 'org.mockito:mockito-core:3.12.4'
+    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.6.0'
+    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine'
+    api project(":${rootProject.name}-core")
```
Comment (joafalves) [inline extensions/ext-tiled/build.gradle]: local api dependences appear first

File: extensions/ext-tiled/build.gradle [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+test {
```
Comment (joafalves) [inline extensions/ext-tiled/build.gradle]: not required, parent gradle already defines this.

File: extensions/ext-tiled/build.gradle [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    useJUnitPlatform()
+}
+
+task unitTest( type: Test ) {
+    exclude 'org/pixel/ext/tiled/**/*Integrated*'
+    useJUnitPlatform()
```
Comment (joafalves) [inline extensions/ext-tiled/build.gradle:31]: not required, parent gradle already defines this `useJUnitPlatform()`.

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledAnimation.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
+package org.pixel.ext.tiled.content;
+
+import java.util.List;
+
+/**
+ * An animation for a TiledTile of a TiledTileSet. Uses an ordered list of TiledFrames to store animation frames.
+ */
+public class TiledAnimation {
+    List<TiledFrame> frameList;
```
Comment (joafalves) [inline extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledAnimation.java]: private?

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledConstants.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
+package org.pixel.ext.tiled.content;
+
+public enum TiledConstants {
+    PIXEL_EPSILON(0.005f);
```
Comment (joafalves) [inline extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledConstants.java:4]: private(s)?

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledFlipMasks.java [hunk 13]
```diff
@@ -0,0 +1,17 @@
+package org.pixel.ext.tiled.content;
+
+public enum TiledFlipMasks {
+    HORIZONTAL_FLIP_FLAG(0x80000000),
+    VERTICAL_FLIP_FLAG(0x40000000),
+    DIAGONAL_FLIP_FLAG(0x20000000);
+
+    long bits;
```
Comment (joafalves) [inline extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledFlipMasks.java]: private?

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledImageLayer.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
+package org.pixel.ext.tiled.content;
+
+import org.pixel.content.Texture;
+import org.pixel.ext.tiled.view.TiledGenericMapView;
+
+/**
+ * A TileMap layer that displays an image.
+ */
+public class TiledImageLayer extends TiledLayer {
+    Texture image;
```
Comment (joafalves) [inline extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledImageLayer.java]: private?

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledLayerGroup.java [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+import org.pixel.ext.tiled.view.TiledGenericMapView;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A group of TiledLayers.
+ */
+public class TiledLayerGroup extends TiledLayer {
+    List<TiledLayer> layers;
```
Comment (joafalves) [inline extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledLayerGroup.java]: private? (check all these)

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledObject.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+        this.position = position;
+    }
+
+    /**
+     * Accepts the TiledGenericObjectGroupView Visitor to draw the object.
+     * @param group The group this object belongs to.
+     * @param view The view visitor that draws the object.
+     */
+    public void draw(TiledObjectGroup group, TiledGenericObjectGroupView view) {
+
```
Comment (joafalves) [inline extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledObject.java]: empty draw?

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledPoint.java [hunk 17]
```diff
@@ -0,0 +1,7 @@
+package org.pixel.ext.tiled.content;
+
+/**
+ * A point TiledObject.
+ */
+public class TiledPoint extends TiledObject {
```
Comment (joafalves) [inline extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledPoint.java:8]: is this being used anywhere?

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledTileLayer.java [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * A TiledTileLayer is a TiledLayer that contains tiles.
+ */
+public class TiledTileLayer extends TiledLayer {
+    private final int width;
+    private final int height;
+    private final long[][] tiles;
+
+    public TiledTileLayer(int width, int height, TiledMap tileMap) {
```
Comment (joafalves) [inline extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledTileLayer.java:20]: Missing constructor JavaDoc (check others that have parameters), otherwise it will generate warnings on the javadoc build (you can run this task and see all warnings)

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/importer/ImageLayerCollector.java [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+import org.w3c.dom.Element;
+
+class ImageLayerCollector extends LayerCollector {
+    @Override
+    TiledLayer collect(TiledMap tileMap, Element tileLayerElement, ImportContext ctx) {
+        TiledImageLayer layer = new TiledImageLayer(collectLayerData(tileMap, tileLayerElement));
+
+        Element image = (Element) tileLayerElement.getElementsByTagName("image").item(0);
+
+        if(image == null) {
```
Comment (joafalves) [inline extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/importer/ImageLayerCollector.java]: bad formatting

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/importer/ObjectCollector.java [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+
+class ObjectCollector {
+    public Pair<Integer, TiledObject> collect(Element objectElement) {
+        CustomPropertiesCollector collector = new CustomPropertiesCollector();
+
+        TiledCustomProperties customProperties = collector.collect(objectElement);
+        TiledObject object;
+
+        long gID;
+
```
Comment (joafalves) [inline extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/importer/ObjectCollector.java]: unnecessary space

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/importer/TiledTileSetImporter.java [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+
+            TiledTile tile = new TiledTile();
+
+            NodeList groups = tileElement.getElementsByTagName("objectgroup");
+
+            if (groups.getLength() > 0) {
+                NodeList objects = ((Element) groups.item(0)).getElementsByTagName("object");
+                HashMap<Integer, TiledObject> map = new HashMap<>();
+
+
```
Comment (joafalves) [inline extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/importer/TiledTileSetImporter.java]: remove this extra line

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/utils/XMLUtils.java [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+
+import javax.xml.XMLConstants;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+public class XMLUtils {
```
Comment (joafalves) [inline extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/utils/XMLUtils.java]: Rename to TiledXmlUtils

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/utils/XMLUtils.java [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+            ByteBuffer bb = ctx.getBuffer();
+
+            byte[] array = new byte[bb.limit()];
+            bb.get(array, 0, bb.limit());
+            bb.flip();
+
+            doc = documentBuilder.parse(new ByteArrayInputStream(array, bb.position(), bb.limit()));
+
+        } catch (ParserConfigurationException | SAXException | IOException e) {
+            e.printStackTrace();
```
Comment (joafalves) [inline extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/utils/XMLUtils.java]: Create a logger and log this exception:

LOG.error("Exception caught!", e);  <- if last element is an exception instance, it is automatically appended.

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/view/TiledGenericMapView.java [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * A generic map view.
+ */
+public interface TiledGenericMapView extends TiledView<TiledMap> {
+    /**
+     * Draws a tile layer.
+     * @param layer The layer to draw.
+     */
+    default void draw(TiledTileLayer layer) {
```
Comment (joafalves) [inline extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/view/TiledGenericMapView.java]: since all implementations have these functions, is there any reason for them to have a default here? (check similar ones)

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/view/TiledImageLayerView.java [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+import org.pixel.math.Boundary;
+import org.pixel.math.Vector2;
+
+/**
+ * A view for a tiled image layer.
+ */
+public class TiledImageLayerView implements TiledView<TiledImageLayer> {
+    private final Boundary imageBoundary;
+    private final Vector2 position;
+    SpriteBatch spriteBatch;
```
Comment (joafalves) [inline extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/view/TiledImageLayerView.java]: private?

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/view/TiledImageLayerView.java [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+    /**
+     * Draws the layer.
+     * @param layer The layer to draw.
+     * @param currentMs The current time in milliseconds.
+     */
+    @Override
+    public void draw(TiledImageLayer layer, long currentMs) {
+        position.set((float) layer.getOffsetX(), (float) layer.getOffsetY());
+
+        if(layer.getImage() == null) {
```
Comment (joafalves) [inline extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/view/TiledImageLayerView.java]: bad formatting. Consider using the IntellIJ code formatter on all module files.

File: extensions/ext-tiled/src/test/java/org/pixel/ext/tiled/content/TileSetTest.java [hunk 27]
```diff
@@ -0,0 +1,156 @@
...
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.pixel.content.Texture;
+import org.pixel.math.Rectangle;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class TileSetTest {
```
Comment (joafalves) [inline extensions/ext-tiled/src/test/java/org/pixel/ext/tiled/content/TileSetTest.java:12]: Not really an issue... cool tests x)

File: gradlew.bat [hunk 28]
```diff
@@ -1,89 +1,89 @@
-@rem
```
Comment (joafalves) [inline gradlew.bat]: discard these gradlew changes (gradle.bat, gradle-wrapper.jar, etc...)

File: settings.gradle [hunk 29]
```diff
@@ -11,10 +11,10 @@ include "modules/commons", "modules/core", "modules/input", "modules/math", "mod
 
 // extensions:
 include "extensions/ext-log4j2", "extensions/ext-ldtk", "extensions/ext-physics", "extensions/ext-gui",   \
-   "extensions/ext-ecs", "extensions/ext-ecs-extra"
+   "extensions/ext-ecs", "extensions/ext-ecs-extra", "extensions/ext-tiled"
```
Comment (joafalves) [inline settings.gradle]: discard this file, I will add the dependencies after the PR.

File: modules/math/src/main/java/org/pixel/math/Rectangle.java [hunk 30]
```diff
@@ -459,18 +460,21 @@ public void set(Rectangle other) {
...
 
     @Override
-    public String toString() {
-        return String.format("%s: [x: %f, y: %f, w: %f, h: %f]",
-                this.getClass().getSimpleName(), getX(), getY(), getWidth(), getHeight());
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof Rectangle)) return false;
+        Rectangle rectangle = (Rectangle) o;
+        return Float.compare(rectangle.x, x) == 0 && Float.compare(rectangle.y, y) == 0 && Float.compare(rectangle.width, width) == 0 && Float.compare(rectangle.height, height) == 0;
```
Comment (joafalves) [inline modules/math/src/main/java/org/pixel/math/Rectangle.java]: Keep the previous equals body (add the this == 0). Float.compare is heaver than checking the values (which in a game-oriented library is important)

File: modules/math/src/main/java/org/pixel/math/Vector2.java [hunk 31]
```diff
@@ -455,13 +456,12 @@ public String toString() {
...
-    public boolean equals(Object obj) {
-        if (obj instanceof Vector2) {
-            return ((Vector2) obj).getX() == this.getX() && ((Vector2) obj).getY() == this.getY();
-        }
-        return false;
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof Vector2)) return false;
+        Vector2 vector2 = (Vector2) o;
+        return Float.compare(vector2.x, x) == 0 && Float.compare(vector2.y, y) == 0;
```
Comment (joafalves) [inline modules/math/src/main/java/org/pixel/math/Vector2.java]: Same here.

File: modules/core/src/main/java/org/pixel/graphics/render/SpriteBatch.java [hunk 32]
```diff
@@ -404,9 +404,9 @@ public void draw(Texture texture, Vector2 position, Rectangle source, Color colo
         spriteData.x = position.getX();
         spriteData.y = position.getY();
         spriteData.width =
-                texture.getWidth() * (source != null ? source.getWidth() / texture.getWidth() * scaleX : scaleX);
+                texture.getWidth() * (source != null ? Math.abs(source.getWidth()) / texture.getWidth() * scaleX : scaleX);
```
Comment (joafalves) [inline modules/core/src/main/java/org/pixel/graphics/render/SpriteBatch.java]: why is this here? it should be added by the draw() caller as it impacts performance here.

**Review by joafalves — CHANGES_REQUESTED** [review]
@Homailot please review & update.

File: extensions/ext-tiled/build.gradle (hunk 10 above)
Comment (Homailot) [reply extensions/ext-tiled/build.gradle:31]: When I remove this line the unitTest task fails to run saying "test events were not received"

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledConstants.java (hunk 12 above)
Comment (Homailot) [reply extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledConstants.java:4]: What should be private here? I can't make the enum package-private because it is used in the view package, but I can make the "value" property private

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledObject.java (hunk 16 above)
Comment (Homailot) [reply extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledObject.java]: The best idea here is probably to make a nullObject for object types not recognized by the importer and make this class abstract

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledPoint.java (hunk 17 above)
Comment (Homailot) [reply extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledPoint.java:8]: In tiled you can create objects that are just points which might be useful for indicating spawn points and other features. This class is used to distinguish between the general object and the concrete point

File: modules/math/src/main/java/org/pixel/math/Rectangle.java (hunk 30 above)
Comment (Homailot) [reply modules/math/src/main/java/org/pixel/math/Rectangle.java]: My bad, I created this equals function because it didn't exist at the time

File: modules/core/src/main/java/org/pixel/graphics/render/SpriteBatch.java (hunk 32 above)
Comment (Homailot) [reply modules/core/src/main/java/org/pixel/graphics/render/SpriteBatch.java]: It seems that when I remove this I can't get tile flipping to work at all. I don't remember why I changed this but here is an example with Math.abs:
image


Without Math.abs
image

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledConstants.java (hunk 12 above)
Comment (joafalves) [reply extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledConstants.java:4]: The value only (the 's' was referring to other similar scenarios)

File: extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledTileLayer.java (hunk 18 above)
Comment (Homailot) [reply extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledTileLayer.java:20]: I can't seem to get any warnings when running the javadoc task despite the HTML files still being created:

```
12:03:54: Executing task 'javadoc'...


> Configure project :
Build Targets: [natives-linux]

> Task :pixel-commons:compileJava UP-TO-DATE
> Task :pixel-math:compileJava UP-TO-DATE
> Task :pixel-input:compileJava UP-TO-DATE
> Task :pixel-pipeline:compileJava UP-TO-DATE
> Task :pixel-core:compileJava UP-TO-DATE
> Task :pixel-ext-tiled:compileJava UP-TO-DATE
> Task :pixel-ext-tiled:processResources NO-SOURCE
> Task :pixel-ext-tiled:classes UP-TO-DATE
> Task :pixel-ext-tiled:javadoc UP-TO-DATE

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.4/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 626ms
7 actionable tasks: 7 up-to-date
12:03:56: Task execution finished 'javadoc'.
```

Comment (Homailot) [reply extensions/ext-tiled/src/main/java/org/pixel/ext/tiled/content/TiledTileLayer.java:20]: Command line version:
```
gradle javadoc --warning-mode all

> Configure project :
Build Targets: [natives-linux]

> Configure project :pixel-learning
The JavaApplication.setMainClassName(String) method has been deprecated. This is scheduled to be removed in Gradle 8.0. Use #getMainClass().set(...) instead. See https://docs.gradle.org/7.4.1/dsl/org.gradle.api.plugins.JavaApplication.html#org.gradle.api.plugins.JavaApplication:mainClass for more details.
        at build_6ydyqtx7set4zf24bwgbzprc4$_run_closure2.doCall(/home/nunoa/Applications/pixel-community/demos/learning/build.gradle:19)
        (Run with --stacktrace to get the full stack trace of this deprecation warning.)

BUILD SUCCESSFUL in 5s
7 actionable tasks: 2 executed, 5 up-to-date
```

File: modules/core/src/main/java/org/pixel/graphics/render/SpriteBatch.java (hunk 32 above)
Comment (Homailot) [reply modules/core/src/main/java/org/pixel/graphics/render/SpriteBatch.java]: Solved by using the display area parameter
